// @renproject/rpc/build/module/v2/pack/pack.js
import { fromBase64 } from "@renproject/utils";
import BigNumber from "bignumber.js";
export var PackPrimitive;
(function (PackPrimitive) {
    PackPrimitive["Bool"] = "bool";
    PackPrimitive["U8"] = "u8";
    PackPrimitive["U16"] = "u16";
    PackPrimitive["U32"] = "u32";
    PackPrimitive["U64"] = "u64";
    PackPrimitive["U128"] = "u128";
    PackPrimitive["U256"] = "u256";
    PackPrimitive["Str"] = "string";
    PackPrimitive["Bytes"] = "bytes";
    PackPrimitive["Bytes32"] = "bytes32";
    PackPrimitive["Bytes65"] = "bytes65";
})(PackPrimitive || (PackPrimitive = {}));
export const isPackStructType = (type) => {
    return (typeof type === "object" &&
        type.struct !== undefined);
};
export const isPackListType = (type) => {
    return (typeof type === "object" && type.list !== undefined);
};
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const unmarshalPackPrimitive = (type, value) => {
    switch (type) {
        // Booleans
        case PackPrimitive.Bool:
            return value;
        // Integers
        case PackPrimitive.U8:
        case PackPrimitive.U16:
        case PackPrimitive.U32:
        case PackPrimitive.U64:
        case PackPrimitive.U128:
        case PackPrimitive.U256:
            return new BigNumber(value);
        // Strings
        case PackPrimitive.Str:
            return value;
        // Bytes
        case PackPrimitive.Bytes:
        case PackPrimitive.Bytes32:
        case PackPrimitive.Bytes65:
            return fromBase64(value);
    }
};
export const unmarshalPackStruct = (type, value) => {
    const struct = {};
    for (const member of type.struct) {
        const keys = Object.keys(member);
        if (keys.length === 0) {
            throw new Error(`Invalid struct member with no entries.`);
        }
        if (keys.length > 1) {
            throw new Error(`Invalid struct member with multiple entries.`);
        }
        const key = Object.keys(member)[0];
        const memberType = member[key];
        if (value && !value.hasOwnProperty(key)) {
            throw new Error(`Missing pack value for key ${key}.`);
        }
        struct[key] = unmarshalPackValue(memberType, value[key]);
    }
    return struct;
};
/**
 * Unmarshals a pack list.
 */
export const unmarshalPackList = (type, value) => {
    return value.map((element) => unmarshalPackValue(type.list, element));
};
export const unmarshalPackValue = (type, value) => {
    if (isPackListType(type)) {
        return unmarshalPackList(type, value);
    }
    else if (isPackStructType(type)) {
        return unmarshalPackStruct(type, value);
    }
    else if (typeof type === "string") {
        if (type === "nil")
            return null;
        return unmarshalPackPrimitive(type, value);
    }
    let valueString;
    try {
        valueString = JSON.stringify(value);
    }
    catch (_error) {
        valueString = String(value);
    }
    if (valueString.length > 20) {
        valueString = `${valueString.slice(0, 17)}...`;
    }
    throw new Error(`Unknown value type ${String(type)}${!type ? ` for value ${valueString}` : ""}.`);
};
export const unmarshalTypedPackValue = ({ t, v }) => {
    return unmarshalPackValue(t, v);
};
