// @renproject/chains-bitcoin/build/main/script/script.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Script = void 0;
const utils_1 = require("@renproject/utils");
const opcodes_1 = require("./opcodes");
const checksum = (hash) => utils_1.sha256(utils_1.sha256(hash)).slice(0, 4);
class Script {
    constructor() {
        this.OP = opcodes_1.Opcode;
        this.addOp = (op) => {
            this.script = Buffer.concat([this.script, Buffer.from([op])]);
            return this;
        };
        this.addData = (data) => {
            this.script = Buffer.concat([
                this.script,
                Buffer.from([data.length]),
                data,
            ]);
            return this;
        };
        this.toBuffer = () => this.script;
        this.toScriptHashOut = () => {
            return new Script()
                .addOp(Script.OP.OP_HASH160)
                .addData(utils_1.hash160(this.toBuffer()))
                .addOp(Script.OP.OP_EQUAL)
                .toBuffer();
        };
        this.toAddress = (prefix) => {
            // Hash
            const hash = utils_1.hash160(this.toBuffer());
            // Prepend prefix
            const hashWithPrefix = Buffer.concat([prefix, hash]);
            // Append checksum
            const hashWithChecksum = Buffer.concat([
                hashWithPrefix,
                checksum(hashWithPrefix),
            ]);
            return hashWithChecksum;
        };
        this.script = Buffer.from([]);
    }
}
exports.Script = Script;
Script.OP = opcodes_1.Opcode;
