// @renproject/ren/build/main/burnAndRelease.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BurnAndRelease = exports.BurnAndReleaseStatus = void 0;
const interfaces_1 = require("@renproject/interfaces");
const utils_1 = require("@renproject/utils");
const bn_js_1 = __importDefault(require("bn.js"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
var BurnAndReleaseStatus;
(function (BurnAndReleaseStatus) {
    BurnAndReleaseStatus["Pending"] = "pending";
    BurnAndReleaseStatus["Burned"] = "burned";
    BurnAndReleaseStatus["Released"] = "released";
    BurnAndReleaseStatus["Reverted"] = "reverted";
})(BurnAndReleaseStatus = exports.BurnAndReleaseStatus || (exports.BurnAndReleaseStatus = {}));
class BurnAndRelease {
    /** @hidden */
    constructor(renVM, params, config = {}) {
        this.validateParams = () => {
            utils_1.assertObject({
                from: "object | undefined",
                to: "object",
                transaction: "any | undefined",
                burnNonce: "string | number | undefined",
                contractCalls: "any[] | undefined",
                asset: "string",
                txHash: "string | undefined",
                nonce: "Buffer | string | undefined",
                tags: "string[] | undefined",
            }, { params: this.params });
            if (this.params.contractCalls) {
                this.params.contractCalls.map((contractCall) => {
                    utils_1.assertType("string", {
                        sendTo: contractCall.sendTo,
                        contractFn: contractCall.contractFn,
                    });
                });
            }
        };
        /** @hidden */
        this._initialize = async () => {
            this._state.renNetwork =
                this._state.renNetwork ||
                    interfaces_1.getRenNetworkDetails(await this.renVM.getNetwork(this._state.selector));
            if (this.params.from && !this.params.from.renNetwork) {
                await this.params.from.initialize(this._state.renNetwork);
            }
            if (!this.params.to.renNetwork) {
                await this.params.to.initialize(this._state.renNetwork);
            }
            const burnPayload = this.params.to.burnPayload &&
                (await this.params.to.burnPayload(this.params.from.burnPayloadConfig));
            const overwriteParams = this.params.from &&
                this.params.from.getBurnParams &&
                (await this.params.from.getBurnParams(this.params.asset, burnPayload));
            this.params = Object.assign(Object.assign({}, overwriteParams), this.params);
            return this;
        };
        /**
         * TODO: Refresh the BurnAndRelease status by checking the status of the
         * mint-chain transaction and the RenVM transaction.
         *
         * ```ts
         * await burnAndRelease.refreshStatus();
         * // > "released"
         * ```
         */
        // eslint-disable-next-line @typescript-eslint/require-await
        this.refreshStatus = async () => {
            return this.status;
        };
        this.confirmationTarget = async () => {
            if (utils_1.isDefined(this._state.targetConfirmations)) {
                return this._state.targetConfirmations;
            }
            let target;
            const getConfirmationTarget = this.renVM.getConfirmationTarget;
            if (getConfirmationTarget) {
                target = await utils_1.retryNTimes(async () => getConfirmationTarget(this._state.selector, this.params.from), 2);
            }
            this._state.targetConfirmations = utils_1.isDefined(target) ? target : 6;
            return this._state.targetConfirmations;
        };
        /**
         * Read a burn reference from an Ethereum transaction - or submit a
         * transaction first if the transaction details have been provided.
         */
        this.burn = () => {
            const promiEvent = interfaces_1.newPromiEvent();
            (async () => {
                if (this.params.txHash) {
                    return this;
                }
                if (!this.params.from) {
                    throw new Error(`Must either provide field \`to\` or field \`txHash\`.`);
                }
                const { asset, transaction, burnNonce, contractCalls } = this.params;
                let findBurnError;
                try {
                    this.burnDetails = await this.params.from.findBurn(asset, promiEvent, transaction, burnNonce);
                }
                catch (error) {
                    findBurnError = error;
                }
                if (!this.burnDetails && contractCalls) {
                    try {
                        this.burnDetails = await this.params.from.submitBurn(asset, promiEvent, contractCalls, { networkDelay: this._state.config.networkDelay });
                    }
                    catch (error) {
                        findBurnError = findBurnError || error;
                    }
                }
                if (!this.burnDetails) {
                    throw findBurnError || new Error(`Must provide burn details.`);
                }
                this.status = BurnAndReleaseStatus.Burned;
                let confirmationTarget = 1;
                let confirmationTargetError = false;
                try {
                    confirmationTarget = await this.confirmationTarget();
                }
                catch (error) {
                    console.error(error);
                    confirmationTargetError = true;
                }
                let currentConfirmations = 0;
                while (currentConfirmations < confirmationTarget) {
                    try {
                        let target;
                        ({ current: currentConfirmations, target } =
                            await this.params.from.transactionConfidence(this.burnDetails.transaction));
                        if (confirmationTargetError) {
                            confirmationTarget = target;
                        }
                        // Eth based chains only emits until 24 confs;
                        // keep emitting so that we can update the UI
                        promiEvent.emit("confirmation", currentConfirmations, confirmationTarget);
                        // Exit early so that we don't have to sleep if confs are met
                        if (currentConfirmations >= confirmationTarget) {
                            break;
                        }
                    }
                    catch (error) {
                        console.error(error);
                    }
                    await utils_1.sleep(10 * utils_1.SECONDS);
                }
                return this;
            })()
                .then(promiEvent.resolve)
                .catch(promiEvent.reject);
            return promiEvent;
        };
        /**
         * `txHash` returns the RenVM transaction hash, which is distinct from the
         * lock or mint chain transaction hashes. It can be used to query the
         * burn-and-release details from RenVM once they've been submitted to it.
         *
         * The RenVM txHash is a URL-base64 string.
         *
         * ```ts
         * burnAndRelease.txHash();
         * // > "QNM87rNDuxx54H7VK7D_NAU0u_mjk09-G25IJZL1QrI"
         * ```
         */
        this.txHash = () => {
            const txHash = this.params.txHash;
            if (txHash) {
                return utils_1.renVMHashToBase64(txHash, this.renVM.version(this._state.selector) >= 2);
            }
            if (!this.params.from) {
                throw new Error(`Must either provide field \`to\` or field \`txHash\`.`);
            }
            if (!this.burnDetails) {
                throw new Error("Must call `burn` before calling `txHash`.");
            }
            if (this.renVM.version(this._state.selector) >= 2 &&
                this.renVM.burnTxHash) {
                const { transaction, amount, to, nonce } = this.burnDetails;
                const payload = Buffer.from([]);
                const pHash = utils_1.generatePHash([], this._state.logger);
                const { txid, txindex } = this.params.from.transactionRPCFormat(transaction, true);
                const nonceBuffer = new bn_js_1.default(nonce.toFixed()).toArrayLike(Buffer, "be", 32);
                const nHash = utils_1.generateNHash(nonceBuffer, txid, txindex, this.renVM.version(this._state.selector) >= 2, this._state.logger);
                const sHash = utils_1.generateSHash(`${this.params.asset}/to${this.params.to.name}`);
                const gHash = utils_1.generateGHash([], utils_1.Ox(this.params.to.addressToBytes(to)), utils_1.Ox(sHash), nonceBuffer, this.renVM.version(this._state.selector) >= 2, this._state.logger);
                return utils_1.toURLBase64(this.renVM.burnTxHash({
                    selector: this._state.selector,
                    gHash,
                    gPubKey: Buffer.from([]),
                    nHash,
                    nonce: nonceBuffer,
                    output: {
                        txid,
                        txindex,
                    },
                    // FIXME: needed until we support different asset decimals between chains
                    amount: this._state.selector === "FIL/fromSolana"
                        ? amount.multipliedBy(10 ** 9).toFixed()
                        : amount.toFixed(),
                    payload,
                    pHash,
                    to: to.toString(),
                }));
            }
            else {
                return utils_1.toBase64(utils_1.generateBurnTxHash(this._state.selector, this.burnDetails.nonce.toFixed(), this._state.logger));
            }
        };
        /**
         * queryTx requests the status of the burn from RenVM.
         */
        this.queryTx = async () => {
            const burnTransaction = await this.renVM.queryMintOrBurn(this._state.selector, utils_1.fromBase64(this.txHash()));
            this._state.queryTxResult = burnTransaction;
            return burnTransaction;
        };
        /**
         * submit queries RenVM for the status of the burn until the funds are
         * released.
         */
        this.release = () => {
            const promiEvent = interfaces_1.newPromiEvent();
            (async () => {
                if (!this.burnDetails && !this.params.txHash) {
                    throw new Error("Must call `burn` before calling `release`.");
                }
                const txHash = this.txHash();
                if (!this.params.txHash && this.burnDetails) {
                    if (!this.params.from) {
                        throw new Error(`Must either provide field \`to\` or field \`txHash\`.`);
                    }
                    if (this.params.tags && this.params.tags.length > 1) {
                        throw new Error("Providing multiple tags is not supported yet.");
                    }
                    const tags = this.params.tags && this.params.tags.length
                        ? [this.params.tags[0]]
                        : [];
                    const { transaction, amount, to, nonce } = this.burnDetails;
                    try {
                        let returnedTxHash;
                        if (this.renVM.version(this._state.selector) >= 2) {
                            utils_1.assertType("string", { to });
                            const payload = Buffer.from([]);
                            const pHash = utils_1.generatePHash([], this._state.logger);
                            const { txid, txindex } = this.params.from.transactionRPCFormat(transaction, true);
                            const nonceBuffer = new bn_js_1.default(nonce.toFixed()).toArrayLike(Buffer, "be", 32);
                            const nHash = utils_1.generateNHash(nonceBuffer, txid, txindex, this.renVM.version(this._state.selector) >= 2, this._state.logger);
                            const sHash = utils_1.generateSHash(`${this.params.asset}/to${this.params.to.name}`);
                            const gHash = utils_1.generateGHash([], utils_1.Ox(this.params.to.addressToBytes(to)), utils_1.Ox(sHash), nonceBuffer, this.renVM.version(this._state.selector) >= 2, this._state.logger);
                            returnedTxHash = utils_1.toURLBase64(await this.renVM.submitBurn({
                                selector: this._state.selector,
                                tags,
                                gHash,
                                gPubKey: Buffer.from([]),
                                nHash,
                                nonce: nonceBuffer,
                                output: {
                                    txid,
                                    txindex,
                                },
                                // FIXME: hack until we support differing decimals between burn + release chains
                                amount: this._state.selector === "FIL/fromSolana"
                                    ? amount.multipliedBy(10 ** 9).toFixed()
                                    : amount.toFixed(),
                                payload,
                                pHash,
                                to: to.toString(),
                                // from v1
                                burnNonce: new bignumber_js_1.default(0),
                            }));
                        }
                        else {
                            returnedTxHash = utils_1.toBase64(await this.renVM.submitBurn({
                                selector: this._state.selector,
                                tags,
                                burnNonce: nonce,
                                // for v2
                                gHash: Buffer.from([]),
                                gPubKey: Buffer.from([]),
                                nHash: Buffer.from([]),
                                nonce: Buffer.from([]),
                                output: { txid: Buffer.from([]), txindex: "" },
                                amount: "",
                                payload: Buffer.from([]),
                                pHash: Buffer.from([]),
                                to: "",
                            }));
                        }
                        if (txHash && txHash !== returnedTxHash) {
                            this._state.logger.warn(`Unexpected txHash returned from RenVM. Received: ${returnedTxHash}, expected: ${txHash}`);
                        }
                    }
                    catch (error) {
                        // TODO: Check against already being submitted.
                        throw error;
                    }
                }
                // const txHash = await this.renVMNetwork.submitTokenFromEthereum(this.params.sendToken, burnNonce);
                promiEvent.emit("txHash", txHash);
                this._state.logger.debug("txHash:", txHash);
                const response = await this.renVM.waitForTX(this._state.selector, utils_1.fromBase64(txHash), (status) => {
                    promiEvent.emit("status", status);
                    this._state.logger.debug("transaction status:", status);
                }, () => promiEvent._isCancelled(), this._state.config.networkDelay);
                if (response.out && response.out.revert !== undefined) {
                    this.status = BurnAndReleaseStatus.Reverted;
                    this.revertReason = response.out.revert.toString();
                    throw new Error(this.revertReason);
                }
                else {
                    this.status = BurnAndReleaseStatus.Released;
                    if (response.out &&
                        this.renVM.version(this._state.selector) >= 2) {
                        let transaction;
                        try {
                            if (response.out.txid) {
                                const txid = response.out.txid;
                                transaction =
                                    await this.params.to.transactionFromRPCFormat(txid, "", true);
                            }
                            else if (response.out.outpoint) {
                                const { hash, index } = response.out.outpoint;
                                transaction =
                                    await this.params.to.transactionFromRPCFormat(hash, index.toFixed(), true);
                            }
                        }
                        catch (error) {
                            this._state.logger.debug(error);
                        }
                        if (transaction) {
                            this.releaseTransaction = transaction;
                            promiEvent.emit("transaction", transaction);
                        }
                    }
                }
                return response;
            })()
                .then(promiEvent.resolve)
                .catch(promiEvent.reject);
            return promiEvent;
        };
        this.params = params;
        this.renVM = renVM;
        this._state = {
            logger: config.logger || interfaces_1.NullLogger,
            selector: this.renVM.selector(this.params),
            targetConfirmations: undefined,
            config: Object.assign(Object.assign({}, config), { networkDelay: config.networkDelay || 15 * utils_1.SECONDS }),
        };
        this.validateParams();
        this.status = this.params.txHash
            ? BurnAndReleaseStatus.Burned
            : BurnAndReleaseStatus.Pending;
        {
            // Debug log
            const restOfParams = __rest(this.params, []);
            this._state.logger.debug("burnAndRelease created:", restOfParams);
        }
    }
}
exports.BurnAndRelease = BurnAndRelease;
