// @renproject/chains-ethereum/build/main/arbitrum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Arbitrum = exports.ArbitrumClass = exports.ArbitrumConfigMap = exports.renArbitrumMainnet = exports.renArbitrumTestnet = void 0;
const interfaces_1 = require("@renproject/interfaces");
const utils_1 = require("@renproject/utils");
const ethereum_1 = require("./ethereum");
const networks_1 = require("./networks");
const utils_2 = require("./utils");
exports.renArbitrumTestnet = {
    name: "Arbitrum Testnet",
    chain: "ArbitrumTestnet",
    isTestnet: true,
    chainLabel: "Arbitrum Testnet",
    networkID: 421611,
    addresses: {
        GatewayRegistry: "0x5eEBf6c199a9Db26dabF621fB8c43D58C62DF2bd",
        BasicAdapter: "0x1156663dFab56A9BAdd844e12eDD69eC96Dd0eFb",
    },
    publicProvider: () => `https://rinkeby.arbitrum.io/rpc`,
    explorer: networks_1.StandardExplorer("https://rinkeby-explorer.arbitrum.io"),
    /** @deprecated Renamed to publicProvider. Will be removed in 3.0.0. */
    infura: "https://rinkeby.arbitrum.io/rpc",
    /** @deprecated Renamed to explorer. Will be removed in 3.0.0. */
    etherscan: "https://rinkeby-explorer.arbitrum.io",
};
exports.renArbitrumMainnet = {
    name: "Arbitrum Mainnet",
    chain: "ArbitrumMainnet",
    isTestnet: false,
    chainLabel: "Arbitrum Mainnet",
    networkID: 42161,
    addresses: {
        GatewayRegistry: "0x21C482f153D0317fe85C60bE1F7fa079019fcEbD",
        BasicAdapter: "0xAC23817f7E9Ec7EB6B7889BDd2b50e04a44470c5",
    },
    publicProvider: () => `https://arb1.arbitrum.io/rpc`,
    explorer: networks_1.StandardExplorer("https://explorer.arbitrum.io"),
    /** @deprecated Renamed to publicProvider. Will be removed in 3.0.0. */
    infura: "https://arb1.arbitrum.io/rpc",
    /** @deprecated Renamed to explorer. Will be removed in 3.0.0. */
    etherscan: "https://explorer.arbitrum.io",
};
exports.ArbitrumConfigMap = {
    [interfaces_1.RenNetwork.Testnet]: exports.renArbitrumTestnet,
    [interfaces_1.RenNetwork.Mainnet]: exports.renArbitrumMainnet,
};
const resolveArbitrumNetwork = (renNetwork) => {
    if (!renNetwork) {
        return exports.ArbitrumConfigMap[interfaces_1.RenNetwork.Mainnet];
    }
    if (renNetwork.addresses) {
        return renNetwork;
    }
    else {
        const details = interfaces_1.getRenNetworkDetails(
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
        renNetwork);
        return details.isTestnet ? exports.renArbitrumTestnet : exports.renArbitrumMainnet;
    }
};
class ArbitrumClass extends ethereum_1.EthereumClass {
    constructor(web3Provider, renNetwork, config = {}) {
        // To be compatible with the Ethereum chain class, the first parameter
        // is a web3Provider and the second the RenVM network.
        super(web3Provider, resolveArbitrumNetwork(renNetwork), config);
        this.chain = ArbitrumClass.chain;
        this.name = ArbitrumClass.chain;
        this.legacyName = undefined;
        this.logRequestLimit = 20000;
        this.configMap = exports.ArbitrumConfigMap;
        this.utils = utils_1.utilsWithChainNetwork(ArbitrumClass.utils, () => this.renNetworkDetails);
        this.initialize = (renNetwork) => {
            this.renNetworkDetails =
                this.renNetworkDetails ||
                    exports.ArbitrumConfigMap[interfaces_1.getRenNetworkDetails(renNetwork).name];
            if (!this.renNetworkDetails) {
                throw new Error(`Unable to set ${this.name} network for RenVM network ${interfaces_1.getRenNetworkDetails(renNetwork).name}. Please provide ${this.name} network details to ${this.name} constructor.`);
            }
            return this;
        };
    }
}
exports.ArbitrumClass = ArbitrumClass;
ArbitrumClass.chain = "Arbitrum";
ArbitrumClass.configMap = exports.ArbitrumConfigMap;
ArbitrumClass.utils = {
    resolveChainNetwork: resolveArbitrumNetwork,
    addressIsValid: utils_2.addressIsValid,
    transactionIsValid: utils_2.transactionIsValid,
    addressExplorerLink: (address, network) => `${(ArbitrumClass.utils.resolveChainNetwork(network) ||
        exports.renArbitrumMainnet).etherscan}/address/${address}`,
    transactionExplorerLink: (transaction, network) => `${(exports.Arbitrum.utils.resolveChainNetwork(network) ||
        exports.renArbitrumMainnet).etherscan}/tx/${transaction || ""}`,
};
// @dev Removes any static fields, except `utils`.
exports.Arbitrum = utils_1.Callable(ArbitrumClass);
