// @renproject/chains-ethereum/build/main/goerli.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Goerli = exports.GoerliClass = exports.GoerliConfigMap = exports.renGoerli = void 0;
const interfaces_1 = require("@renproject/interfaces");
const utils_1 = require("@renproject/utils");
const ethereum_1 = require("./ethereum");
const networks_1 = require("./networks");
const utils_2 = require("./utils");
exports.renGoerli = {
    name: "Görli Testnet",
    chain: "goerliTestnet",
    isTestnet: true,
    chainLabel: "Görli Testnet",
    networkID: 6284,
    addresses: {
        GatewayRegistry: "0xD881213F5ABF783d93220e6bD3Cc21706A8dc1fC",
        BasicAdapter: "0xD087b0540e172553c12DEEeCDEf3dFD21Ec02066",
    },
    publicProvider: ({ infura } = {}) => `https://goerli.infura.io/v3/${infura || ""}`,
    explorer: networks_1.StandardExplorer("https://goerli.etherscan.io"),
    /** @deprecated Renamed to publicProvider. Will be removed in 3.0.0. */
    infura: "https://goerli.infura.io",
    /** @deprecated Renamed to explorer. Will be removed in 3.0.0. */
    etherscan: "https://goerli.etherscan.io",
};
exports.GoerliConfigMap = {
    [interfaces_1.RenNetwork.Testnet]: exports.renGoerli,
};
const resolveGoerliNetwork = (renNetwork) => {
    if (!renNetwork) {
        return exports.renGoerli;
    }
    if (renNetwork.addresses) {
        return renNetwork;
    }
    else {
        const details = interfaces_1.getRenNetworkDetails(
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
        renNetwork);
        if (!details.isTestnet) {
            throw new Error(`Goerli not supported on mainnet.`);
        }
        return exports.renGoerli;
    }
};
class GoerliClass extends ethereum_1.EthereumClass {
    constructor(web3Provider, renNetwork, config = {}) {
        // To be compatible with the Ethereum chain class, the first parameter
        // is a web3Provider and the second the RenVM network.
        super(web3Provider, resolveGoerliNetwork(renNetwork), config);
        this.chain = GoerliClass.chain;
        this.name = GoerliClass.chain;
        this.legacyName = undefined;
        this.configMap = exports.GoerliConfigMap;
        this.utils = utils_1.utilsWithChainNetwork(GoerliClass.utils, () => this.renNetworkDetails);
        this.initialize = (renNetwork) => {
            this.renNetworkDetails =
                this.renNetworkDetails ||
                    exports.GoerliConfigMap[interfaces_1.getRenNetworkDetails(renNetwork).name];
            if (!this.renNetworkDetails) {
                throw new Error(`Unable to set ${this.name} network for RenVM network ${interfaces_1.getRenNetworkDetails(renNetwork).name}. Please provide ${this.name} network details to ${this.name} constructor.`);
            }
            return this;
        };
    }
}
exports.GoerliClass = GoerliClass;
GoerliClass.chain = "Goerli";
// public logRequestLimit = 1000;
GoerliClass.configMap = exports.GoerliConfigMap;
GoerliClass.utils = {
    resolveChainNetwork: resolveGoerliNetwork,
    addressIsValid: utils_2.addressIsValid,
    transactionIsValid: utils_2.transactionIsValid,
    addressExplorerLink: (address, network) => `${(exports.Goerli.utils.resolveChainNetwork(network) || exports.renGoerli)
        .etherscan}/address/${address}`,
    transactionExplorerLink: (transaction, network) => `${(exports.Goerli.utils.resolveChainNetwork(network) || exports.renGoerli)
        .etherscan}/tx/${transaction || ""}`,
};
// @dev Removes any static fields, except `utils`.
exports.Goerli = utils_1.Callable(GoerliClass);
