// @renproject/chains-ethereum/build/main/polygon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Polygon = exports.PolygonClass = exports.PolygonConfigMap = exports.renPolygonMainnet = exports.renPolygonTestnet = void 0;
const interfaces_1 = require("@renproject/interfaces");
const utils_1 = require("@renproject/utils");
const ethereum_1 = require("./ethereum");
const networks_1 = require("./networks");
const utils_2 = require("./utils");
exports.renPolygonTestnet = {
    name: "Polygon Testnet",
    chain: "polygonTestnet",
    isTestnet: true,
    chainLabel: "Polygon Testnet",
    networkID: 80001,
    addresses: {
        GatewayRegistry: "0xD881213F5ABF783d93220e6bD3Cc21706A8dc1fC",
        BasicAdapter: "0xD087b0540e172553c12DEEeCDEf3dFD21Ec02066",
    },
    publicProvider: () => `https://rpc-mumbai.maticvigil.com`,
    explorer: networks_1.StandardExplorer("https://mumbai.polygonscan.com"),
    /** @deprecated Renamed to publicProvider. Will be removed in 3.0.0. */
    infura: "https://rpc-mumbai.maticvigil.com",
    /** @deprecated Renamed to explorer. Will be removed in 3.0.0. */
    etherscan: "https://mumbai.polygonscan.com/",
};
exports.renPolygonMainnet = {
    name: "Polygon Mainnet",
    chain: "polygonMainnet",
    isTestnet: false,
    chainLabel: "Polygon Mainnet",
    networkID: 137,
    addresses: {
        GatewayRegistry: "0x21C482f153D0317fe85C60bE1F7fa079019fcEbD",
        BasicAdapter: "0xAC23817f7E9Ec7EB6B7889BDd2b50e04a44470c5",
    },
    publicProvider: () => `https://rpc-mainnet.maticvigil.com`,
    explorer: networks_1.StandardExplorer("https://polygonscan.com"),
    /** @deprecated Renamed to publicProvider. Will be removed in 3.0.0. */
    infura: "https://rpc-mainnet.maticvigil.com",
    /** @deprecated Renamed to explorer. Will be removed in 3.0.0. */
    etherscan: "https://polygonscan.com",
};
exports.PolygonConfigMap = {
    [interfaces_1.RenNetwork.Testnet]: exports.renPolygonTestnet,
    [interfaces_1.RenNetwork.Mainnet]: exports.renPolygonMainnet,
};
const resolvePolygonNetwork = (renNetwork) => {
    if (!renNetwork) {
        return exports.PolygonConfigMap[interfaces_1.RenNetwork.Mainnet];
    }
    if (renNetwork.addresses) {
        return renNetwork;
    }
    else {
        const details = interfaces_1.getRenNetworkDetails(
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
        renNetwork);
        return details.isTestnet ? exports.renPolygonTestnet : exports.renPolygonMainnet;
    }
};
class PolygonClass extends ethereum_1.EthereumClass {
    constructor(web3Provider, renNetwork, config = {}) {
        // To be compatible with the Ethereum chain class, the first parameter
        // is a web3Provider and the second the RenVM network.
        super(web3Provider, resolvePolygonNetwork(renNetwork), config);
        this.chain = PolygonClass.chain;
        this.name = PolygonClass.chain;
        this.legacyName = undefined;
        this.logRequestLimit = 1000;
        this.configMap = exports.PolygonConfigMap;
        this.utils = utils_1.utilsWithChainNetwork(PolygonClass.utils, () => this.renNetworkDetails);
        this.initialize = (renNetwork) => {
            this.renNetworkDetails =
                this.renNetworkDetails ||
                    exports.PolygonConfigMap[interfaces_1.getRenNetworkDetails(renNetwork).name];
            if (!this.renNetworkDetails) {
                throw new Error(`Unable to set ${this.name} network for RenVM network ${interfaces_1.getRenNetworkDetails(renNetwork).name}. Please provide ${this.name} network details to ${this.name} constructor.`);
            }
            return this;
        };
    }
}
exports.PolygonClass = PolygonClass;
PolygonClass.chain = "Polygon";
PolygonClass.configMap = exports.PolygonConfigMap;
PolygonClass.utils = {
    resolveChainNetwork: resolvePolygonNetwork,
    addressIsValid: utils_2.addressIsValid,
    transactionIsValid: utils_2.transactionIsValid,
    addressExplorerLink: (address, network) => `${(exports.Polygon.utils.resolveChainNetwork(network) ||
        exports.renPolygonMainnet).etherscan}/address/${address}`,
    transactionExplorerLink: (transaction, network) => `${(exports.Polygon.utils.resolveChainNetwork(network) ||
        exports.renPolygonMainnet).etherscan}/tx/${transaction || ""}`,
};
// @dev Removes any static fields, except `utils`.
exports.Polygon = utils_1.Callable(PolygonClass);
