// @renproject/chains-solana/build/main/layouts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GatewayRegistryStateKey = exports.GatewayRegistryLayout = exports.GatewayStateKey = exports.GatewayLayout = exports.RenVmMsgLayout = exports.MintLogLayout = exports.BurnLogLayout = void 0;
const borsh_1 = require("@project-serum/borsh");
exports.BurnLogLayout = borsh_1.struct([
    // amount's type is `spl_math::uint::U256`, which borsh doesn't support.
    borsh_1.u64("amount_section_1"),
    borsh_1.u64("amount_section_2"),
    borsh_1.u64("amount_section_3"),
    borsh_1.u64("amount_section_4"),
    borsh_1.u8("recipient_len"),
    borsh_1.array(borsh_1.u8(), 32, "recipient"),
]);
exports.MintLogLayout = borsh_1.struct([borsh_1.bool("is_initialized")]);
exports.RenVmMsgLayout = borsh_1.struct([
    borsh_1.array(borsh_1.u8(), 32, "p_hash"),
    borsh_1.array(borsh_1.u8(), 32, "amount"),
    borsh_1.array(borsh_1.u8(), 32, "token"),
    borsh_1.array(borsh_1.u8(), 32, "to"),
    borsh_1.array(borsh_1.u8(), 32, "n_hash"),
]);
exports.GatewayLayout = borsh_1.struct([
    borsh_1.bool("is_initialized"),
    borsh_1.array(borsh_1.u8(), 20, "renvm_authority"),
    borsh_1.array(borsh_1.u8(), 32, "selectors"),
    borsh_1.u64("burn_count"),
    borsh_1.u8("underlying_decimals"),
]);
exports.GatewayStateKey = "GatewayStateV0.1.4";
exports.GatewayRegistryLayout = borsh_1.struct([
    borsh_1.bool("is_initialized"),
    borsh_1.publicKey("owner"),
    borsh_1.u64("count"),
    borsh_1.vec(borsh_1.array(borsh_1.u8(), 32), "selectors"),
    borsh_1.vec(borsh_1.publicKey(), "gateways"),
]);
exports.GatewayRegistryStateKey = "GatewayRegistryState";
