// @renproject/chains-solana/build/main/networks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.renLocalnet = exports.renDevnet = exports.renTestnet = exports.renMainnet = exports.resolveNetwork = void 0;
const interfaces_1 = require("@renproject/interfaces");
const isSolNetworkConfig = (x) => {
    return x.genesisHash !== undefined;
};
const resolveNetwork = (renNetwork) => {
    let networkString = "";
    if (typeof renNetwork !== "string") {
        if (isSolNetworkConfig(renNetwork)) {
            return renNetwork;
        }
        else {
            networkString = renNetwork.name;
        }
    }
    else {
        networkString = renNetwork;
    }
    switch (networkString) {
        case interfaces_1.RenNetwork.Mainnet:
            return exports.renMainnet;
        case interfaces_1.RenNetwork.Testnet:
            return exports.renTestnet;
        case interfaces_1.RenNetwork.Devnet:
            return exports.renDevnet;
    }
    return exports.renLocalnet;
};
exports.resolveNetwork = resolveNetwork;
exports.renMainnet = {
    name: interfaces_1.RenNetwork.Mainnet,
    chain: "mainnet",
    isTestnet: false,
    chainLabel: "Mainnet",
    endpoint: "https://ren.rpcpool.com/",
    chainExplorer: "https://explorer.solana.com",
    lightnode: "https://lightnode-mainnet.herokuapp.com",
    addresses: {
        GatewayRegistry: "REGrPFKQhRneFFdUV3e9UDdzqUJyS6SKj88GdXFCRd2",
    },
    genesisHash: "5eykt4UsFv8P8NJdTREpY1vzqKqZKvdpKuc147dw2N9d",
};
exports.renTestnet = {
    name: interfaces_1.RenNetwork.Testnet,
    chain: "testnet",
    isTestnet: true,
    chainLabel: "Testnet",
    endpoint: "https://api.devnet.solana.com",
    chainExplorer: "https://explorer.solana.com",
    lightnode: "https://lightnode-testnet.herokuapp.com",
    addresses: {
        GatewayRegistry: "REGrPFKQhRneFFdUV3e9UDdzqUJyS6SKj88GdXFCRd2",
    },
    genesisHash: "EtWTRABZaYq6iMfeYKouRu166VU2xqa1wcaWoxPkrZBG",
};
exports.renDevnet = {
    name: interfaces_1.RenNetwork.Devnet,
    chain: "devnet",
    isTestnet: true,
    chainLabel: "Devnet",
    endpoint: "https://api.testnet.solana.com",
    chainExplorer: "https://explorer.solana.com",
    lightnode: "https://lightnode-devnet.herokuapp.com",
    addresses: {
        GatewayRegistry: "REGrPFKQhRneFFdUV3e9UDdzqUJyS6SKj88GdXFCRd2",
    },
    genesisHash: "4uhcVJyU9pJkvQyS88uRDiswHXSCkY3zQawwpjk2NsNY",
};
exports.renLocalnet = {
    name: interfaces_1.RenNetwork.Localnet,
    chain: "localnet",
    isTestnet: true,
    chainLabel: "",
    endpoint: "http://0.0.0.0:8899",
    lightnode: "http://0.0.0.0:5000",
    chainExplorer: "https://explorer.solana.com",
    addresses: {
        GatewayRegistry: "REGrPFKQhRneFFdUV3e9UDdzqUJyS6SKj88GdXFCRd2",
    },
    genesisHash: "",
};
