// @renproject/chains-bitcoin/build/main/dogecoin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dogecoin = exports.DogecoinClass = void 0;
const utils_1 = require("@renproject/utils");
const bs58_1 = __importDefault(require("bs58"));
const blockchair_1 = require("./APIs/blockchair");
const sochain_1 = require("./APIs/sochain");
const bitcoin_1 = require("./bitcoin");
const utils_2 = require("./utils");
class DogecoinClass extends bitcoin_1.BitcoinClass {
    constructor() {
        super(...arguments);
        this.chain = DogecoinClass.chain;
        this.name = DogecoinClass.chain;
        this.legacyName = undefined;
        // APIs
        this.withDefaultAPIs = (network) => {
            switch (network) {
                case "mainnet":
                    // prettier-ignore
                    return this
                        .withAPI(blockchair_1.Blockchair(blockchair_1.BlockchairNetwork.DOGECOIN))
                        .withAPI(sochain_1.SoChain(sochain_1.SoChainNetwork.DOGE), { priority: 15 });
                case "testnet":
                    // prettier-ignore
                    return this
                        .withAPI(sochain_1.SoChain(sochain_1.SoChainNetwork.DOGETEST), { priority: 15 });
                case "regtest":
                    throw new Error(`Regtest is currently not supported.`);
            }
        };
        this.asset = "DOGE";
        this.utils = utils_1.utilsWithChainNetwork(DogecoinClass.utils, () => this.chainNetwork);
    }
}
exports.DogecoinClass = DogecoinClass;
DogecoinClass.chain = "Dogecoin";
DogecoinClass.asset = "DOGE";
DogecoinClass.utils = {
    resolveChainNetwork: bitcoin_1.BitcoinClass.utils.resolveChainNetwork,
    p2shPrefix: {
        mainnet: Buffer.from([0x16]),
        testnet: Buffer.from([0xc4]),
    },
    addressBufferToString: bs58_1.default.encode,
    addressIsValid: (address, network = "mainnet") => utils_2.validateAddress(address, DogecoinClass.asset, exports.Dogecoin.utils.resolveChainNetwork(network)),
    transactionIsValid: (transaction, _network = "mainnet") => utils_1.isHex(typeof transaction === "string"
        ? transaction
        : transaction.txHash, { length: 32 }),
    addressExplorerLink: (address, network = "mainnet") => {
        switch (exports.Dogecoin.utils.resolveChainNetwork(network)) {
            case "mainnet":
                return `https://sochain.com/address/DOGE/${address}/`;
            case "testnet":
                return `https://sochain.com/address/DOGETEST/${address}/`;
            case "regtest":
                return undefined;
        }
    },
    transactionExplorerLink: (tx, network = "mainnet") => {
        const txHash = typeof tx === "string" ? tx : tx.txHash;
        switch (exports.Dogecoin.utils.resolveChainNetwork(network)) {
            case "mainnet":
                return `https://sochain.com/tx/DOGE/${txHash}/`;
            case "testnet":
                return `https://sochain.com/tx/DOGETEST/${txHash}/`;
            case "regtest":
                return undefined;
        }
    },
};
exports.Dogecoin = utils_1.Callable(DogecoinClass);
const _ = exports.Dogecoin;
