// @renproject/chains-bitcoin/build/main/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateAddress = void 0;
const wallet_address_validator_1 = require("wallet-address-validator");
const bitcoin_validator_1 = __importDefault(require("wallet-address-validator/src/bitcoin_validator"));
const validateAddress = (address, asset, network) => {
    if (asset === "DGB") {
        const currency = {
            name: "digibyte",
            symbol: "dgb",
            addressTypes: { prod: ["1e", "3f"], testnet: ["7e", "8c"] },
            validator: bitcoin_validator_1.default,
            segwitHrp: network === "mainnet" ? "dgb" : "dgbt",
        };
        return currency.validator.isValidAddress(address, currency, network);
    }
    return wallet_address_validator_1.validate(address, asset, network === "mainnet" ? "prod" : network);
};
exports.validateAddress = validateAddress;
