// @renproject/interfaces/build/module/networks.js
export var RenNetwork;
(function (RenNetwork) {
    RenNetwork["Mainnet"] = "mainnet";
    RenNetwork["Testnet"] = "testnet";
    RenNetwork["Devnet"] = "devnet";
    RenNetwork["Localnet"] = "localnet";
    // Staging
    /**
     * @deprecated Replaced by Mainnet - will be removed in v3
     */
    RenNetwork["MainnetVDot3"] = "mainnet";
    /**
     * @deprecated Replaced by Testnet - will be removed in v3
     */
    RenNetwork["TestnetVDot3"] = "testnet";
    /**
     * @deprecated Replaced by Devnet - will be removed in v3
     */
    RenNetwork["DevnetVDot3"] = "devnet";
})(RenNetwork || (RenNetwork = {}));
const renMainnet = {
    name: RenNetwork.Mainnet,
    lightnode: "https://lightnode-mainnet.herokuapp.com",
    isTestnet: false,
};
const renTestnet = {
    name: RenNetwork.Testnet,
    lightnode: "https://lightnode-testnet.herokuapp.com",
    isTestnet: true,
};
const renDevnet = {
    name: RenNetwork.Devnet,
    lightnode: "https://lightnode-devnet.herokuapp.com/",
    isTestnet: true,
};
const renLocalnet = {
    name: RenNetwork.Localnet,
    lightnode: "http://localhost:5000",
    isTestnet: true,
};
export const getRenNetworkDetails = (renNetwork) => {
    switch (renNetwork) {
        case RenNetwork.Mainnet:
        case RenNetwork.Mainnet:
        case "mainnet-v0.3":
            return renMainnet;
        case RenNetwork.Testnet:
        case RenNetwork.Testnet:
        case "testnet-v0.3":
            return renTestnet;
        case RenNetwork.Devnet:
        case RenNetwork.Devnet:
        case "devnet-v0.3":
            return renDevnet;
        case RenNetwork.Localnet:
        case RenNetwork.Localnet:
            return renLocalnet;
        default:
            return renNetwork;
    }
};
export const RenNetworks = [
    RenNetwork.Mainnet,
    RenNetwork.Testnet,
    RenNetwork.Devnet,
    RenNetwork.Localnet,
];
