// @renproject/ren/build/module/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { LogLevel, RenNetwork, SimpleLogger, } from "@renproject/interfaces";
import { CombinedProvider } from "@renproject/rpc";
import { RenVMProvider } from "@renproject/rpc/build/main/v2";
import { fromSmallestUnit, randomNonce, toSmallestUnit, } from "@renproject/utils";
import { BurnAndRelease } from "./burnAndRelease";
import { defaultDepositHandler } from "./defaultDepositHandler";
import { LockAndMint, DepositStatus, LockAndMintDeposit } from "./lockAndMint";
export { BurnAndRelease } from "./burnAndRelease";
export { LockAndMint, DepositStatus, LockAndMintDeposit } from "./lockAndMint";
/**
 * This is the main exported class from `@renproject/ren`.
 *
 * ```typescript
 * import RenJS from "@renproject/ren";
 * ```
 *
 * By default, RenJS will connect to the RenVM mainnet network. To connect
 * to `testnet` or to configure a custom connection, RenJS takes an optional
 * provider object. See the [[constructor]] for more details.
 *
 * ```typescript
 * new RenJS(); // Same as `new RenJS("mainnet");`
 * new RenJS("testnet");
 * new RenJS(custom provider object);
 * ```
 *
 * It then exposes two main functions:
 * 1. [[lockAndMint]] - for transferring assets to Ethereum.
 * 2. [[burnAndRelease]] - for transferring assets out of Ethereum.
 *
 * Also see:
 * 1. [[getFees]] - for estimating the fees that will be incurred by minting or
 * burning.
 * 2. [[defaultDepositHandler]]
 *
 */
export default class RenJS {
    /**
     * Accepts the name of a network, or a network object.
     *
     * @param network Provide the name of a network - `"mainnet"` or `"testnet"` - or a network object.
     * @param providerOrConfig Provide a custom RPC provider, or provide RenJS configuration settings.
     */
    constructor(providerOrNetwork, config) {
        // const provider: string | Provider | undefined;
        // let config: RenJSConfig | undefined;
        // if (
        //     providerOrConfig &&
        //     (typeof providerOrConfig === "string" ||
        //         (providerOrConfig as Provider).sendMessage)
        // ) {
        //     provider = providerOrConfig as string | Provider;
        // } else if (providerOrConfig) {
        //     config = providerOrConfig as RenJSConfig;
        // }
        /**
         * @hidden
         */
        this.utils = RenJS.utils;
        this.getFees = ({ asset, from, to, }) => __awaiter(this, void 0, void 0, function* () {
            if (!(yield from.assetIsSupported(asset))) {
                throw new Error(`Asset not supported by chain ${from.name}.`);
            }
            if (!(yield to.assetIsSupported(asset))) {
                throw new Error(`Asset not supported by chain ${to.name}.`);
            }
            if (yield from.assetIsNative(asset)) {
                // LockAndMint
                return yield this.renVM.estimateTransactionFee(asset, from, to);
            }
            else if (yield to.assetIsNative(asset)) {
                // BurnAndRelease
                return yield this.renVM.estimateTransactionFee(asset, to, from);
            }
            else {
                // BurnAndMint
                return yield this.renVM.estimateTransactionFee(asset, from, to);
            }
        });
        /**
         * `lockAndMint` initiates the process of bridging an asset from its native
         * chain to a host chain.
         *
         * See [[LockAndMintParams]] for all the options that can be set.
         *
         * Returns a [[LockAndMint]] object.
         *
         * Example initialization:
         *
         * ```js
         * const lockAndMint = renJS.lockAndMint({
         *     asset: "BTC",
         *     from: Bitcoin(),
         *     to: Ethereum(web3Provider).Account({
         *         address: "0x...",
         *     }),
         * });
         * ```
         *
         * @param params See [[LockAndMintParams]].
         */
        this.lockAndMint = (params, config) => __awaiter(this, void 0, void 0, function* () {
            return new LockAndMint(this.renVM, params, Object.assign(Object.assign({}, this._config), config))._initialize();
        });
        /**
         * `burnAndRelease` submits a burn log to RenVM.
         * Returns a [[BurnAndRelease]] object.
         */
        this.burnAndRelease = (params, config) => __awaiter(this, void 0, void 0, function* () {
            return new BurnAndRelease(this.renVM, params, Object.assign(Object.assign({}, this._config), config))._initialize();
        });
        this._config = config || {};
        this._logger =
            (config && config.logger) ||
                new SimpleLogger((config && config.logLevel) || LogLevel.Error);
        this._config.logger = this._logger;
        const defaultProvider = () => config && config.useV2TransactionFormat
            ? new RenVMProvider(
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
            (providerOrNetwork || RenNetwork.Mainnet), undefined, this._logger)
            : new CombinedProvider(
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
            (providerOrNetwork || RenNetwork.Mainnet), this._logger);
        // Use provided provider, provider URL or default lightnode URL.
        this.renVM =
            providerOrNetwork &&
                typeof providerOrNetwork !== "string" &&
                providerOrNetwork.sendMessage
                ? providerOrNetwork
                : defaultProvider();
    }
}
// /**
//  * [STATIC] `Tokens` exposes the tokens that can be passed in to the lockAndMint and
//  * burnAndRelease methods.
//  */
// public static Tokens = Tokens;
/**
 * `Networks` exposes the network options that can be passed in to the RenJS
 * constructor. `Networks.Mainnet` resolves to the string `"mainnet"`.
 */
RenJS.Networks = RenNetwork;
/**
 * A collection of helper functions. [[utils.randomNonce]] can be be used to
 * generate a nonce when calling [[RenJS.lockAndMint]].
 */
RenJS.utils = {
    randomNonce,
    toSmallestUnit,
    fromSmallestUnit,
    fromAscii: (str) => Buffer.from(str),
};
/**
 * `RenJS.defaultDepositHandler` can be passed as a deposit callback when
 * minting. It will handle submitting to RenVM and then to the mint-chain,
 * as long as a valid provider for the mint-chain is given.
 *
 * This is not recommended for front-ends, since it may trigger a wallet
 * pop-up unexpectedly when the mint is ready to be submitted.
 *
 * ```ts
 * lockAndMint.on("deposit", RenJS.defaultDepositHandler);
 * ```
 */
RenJS.defaultDepositHandler = defaultDepositHandler;
// ////////////////////////////////////////////////////////////////////////// //
// EXPORTS                                                                    //
// Based on https://github.com/MikeMcl/bignumber.js/blob/master/bignumber.js  //
// ////////////////////////////////////////////////////////////////////////// //
/* eslint-disable @typescript-eslint/ban-ts-comment, @typescript-eslint/no-explicit-any */
RenJS.default = RenJS.RenJS = RenJS;
RenJS.LockAndMint = LockAndMint;
RenJS.BurnAndRelease = BurnAndRelease;
RenJS.DepositStatus = DepositStatus;
RenJS.LockAndMintDeposit = LockAndMintDeposit;
// AMD
try {
    // @ts-ignore
    if (typeof define === "function" && define.amd) {
        // @ts-ignore
        define(() => RenJS);
    }
}
catch (error) {
    /* ignore */
}
// Node.js and other environments that support module.exports.
try {
    // @ts-ignore
    if (typeof module !== "undefined" && module.exports) {
        module.exports = RenJS;
    }
}
catch (error) {
    /* ignore */
}
// Browser.
try {
    // @ts-ignore
    if (typeof window !== "undefined" && window) {
        window.RenJS = RenJS;
    }
}
catch (error) {
    /* ignore */
}
