// @renproject/gateway-sol/typechain/factories/MintGatewayStateV2__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  MintGatewayStateV2,
  MintGatewayStateV2Interface,
} from "../MintGatewayStateV2";

const _abi = [
  {
    constant: true,
    inputs: [],
    name: "selectorHash",
    outputs: [
      {
        internalType: "bytes32",
        name: "",
        type: "bytes32",
      },
    ],
    payable: false,
    stateMutability: "view",
    type: "function",
  },
];

const _bytecode =
  "0x6080604052348015600f57600080fd5b5060848061001e6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c80632976297614602d575b600080fd5b60336049565b6040518082815260200191505060405180910390f35b6001548156fea265627a7a72315820f554993787802ce6dc14968c46f8288e4a6a1045694db0386e78f094d976f05664736f6c63430005110032";

export class MintGatewayStateV2__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<MintGatewayStateV2> {
    return super.deploy(overrides || {}) as Promise<MintGatewayStateV2>;
  }
  getDeployTransaction(
    overrides?: Overrides & { from?: string | Promise<string> }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): MintGatewayStateV2 {
    return super.attach(address) as MintGatewayStateV2;
  }
  connect(signer: Signer): MintGatewayStateV2__factory {
    return super.connect(signer) as MintGatewayStateV2__factory;
  }
  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): MintGatewayStateV2Interface {
    return new utils.Interface(_abi) as MintGatewayStateV2Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): MintGatewayStateV2 {
    return new Contract(address, _abi, signerOrProvider) as MintGatewayStateV2;
  }
}
