// @renproject/chains-bitcoin/build/main/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BitcoinBaseChain = void 0;
const bech32_1 = __importDefault(require("bech32"));
const interfaces_1 = require("@renproject/interfaces");
const utils_1 = require("@renproject/utils");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const bs58_1 = __importDefault(require("bs58"));
const API_1 = require("./APIs/API");
const blockchair_1 = require("./APIs/blockchair");
const blockstream_1 = require("./APIs/blockstream");
const sochain_1 = require("./APIs/sochain");
const index_1 = require("./script/index");
const transactionToDeposit = (transaction) => ({
    transaction,
    amount: transaction.amount.toString(),
});
/**
 * A base Bitcoin chain class that is extended by each Bitcoin chain/fork.
 */
class BitcoinBaseChain {
    constructor(network) {
        this.chain = BitcoinBaseChain.chain;
        this.name = BitcoinBaseChain.chain;
        this.legacyName = "Btc";
        this.asset = "BTC";
        // APIs
        this.withDefaultAPIs = (network) => {
            switch (network) {
                case "mainnet":
                    // prettier-ignore
                    return this
                        .withAPI(blockstream_1.Blockstream())
                        .withAPI(blockchair_1.Blockchair())
                        .withAPI(sochain_1.SoChain(), { priority: 15 });
                case "testnet":
                    // prettier-ignore
                    return this
                        .withAPI(blockstream_1.Blockstream({ testnet: true }))
                        .withAPI(blockchair_1.Blockchair(blockchair_1.BlockchairNetwork.BITCOIN_TESTNET))
                        .withAPI(sochain_1.SoChain(sochain_1.SoChainNetwork.BTCTEST), { priority: 15 });
                case "regtest":
                    // Will be supported when Electrum is added as an API.
                    throw new Error(`Regtest is currently not supported.`);
            }
        };
        this.api = API_1.CombinedAPI();
        this.withAPI = (api, { priority = 0 } = {}) => {
            this.api.withAPI(api, { priority });
            return this;
        };
        this.utils = utils_1.utilsWithChainNetwork(BitcoinBaseChain.utils, () => this.chainNetwork);
        /**
         * See [[LockChain.initialize]].
         */
        this.initialize = (renNetwork) => {
            this.renNetwork = interfaces_1.getRenNetworkDetails(renNetwork);
            // Prioritize the network passed in to the constructor.
            this.chainNetwork =
                this.chainNetwork ||
                    (this.renNetwork.isTestnet ? "testnet" : "mainnet");
            return this.withDefaultAPIs(this.chainNetwork);
        };
        /**
         * See [[LockChain.assetIsNative]].
         */
        this.assetIsNative = (asset) => asset === this.asset;
        this.assetIsSupported = this.assetIsNative;
        this.assertAssetIsSupported = (asset) => {
            if (!this.assetIsNative(asset)) {
                throw new Error(`Asset ${asset} not supported on ${this.chain}.`);
            }
        };
        /**
         * See [[LockChain.assetDecimals]].
         */
        this.assetDecimals = (asset) => {
            if (asset === this.asset) {
                return 8;
            }
            throw new Error(`Asset ${asset} not supported on ${this.chain}.`);
        };
        /**
         * See [[LockChain.getDeposits]].
         */
        this.getDeposits = async (asset, address, progress, onDeposit) => {
            if (!this.chainNetwork) {
                throw new Error(`${this.name} object not initialized`);
            }
            if (this.chainNetwork === "regtest") {
                throw new Error(`Unable to fetch deposits on ${this.chainNetwork}`);
            }
            this.assertAssetIsSupported(asset);
            let txs;
            if (!progress) {
                try {
                    txs = await utils_1.retryNTimes(() => this.api.fetchTXs(address), 2);
                }
                catch (error) {
                    // Ignore error and fallback to getUTXOs.
                }
            }
            if (!txs) {
                txs = await this.api.fetchUTXOs(address);
            }
            await Promise.all(txs.map(async (tx) => onDeposit(transactionToDeposit(tx))));
            return true;
        };
        /**
         * See [[LockChain.transactionConfidence]].
         */
        this.transactionConfidence = async (transaction) => {
            if (!this.chainNetwork) {
                throw new Error(`${this.name} object not initialized`);
            }
            transaction = await this.api.fetchUTXO(transaction.txHash, transaction.vOut);
            return {
                current: transaction.confirmations,
                target: this.chainNetwork === "mainnet" ? 6 : 2,
            };
        };
        /**
         * See [[LockChain.getGatewayAddress]].
         */
        this.getGatewayAddress = (asset, publicKey, gHash) => {
            if (!this.chainNetwork) {
                throw new Error(`${this.name} object not initialized`);
            }
            this.assertAssetIsSupported(asset);
            const isTestnet = this.chainNetwork === "testnet";
            return this.utils.addressBufferToString(index_1.createAddressBuffer(utils_1.hash160(publicKey), gHash, this.utils.p2shPrefix[isTestnet ? "testnet" : "mainnet"]));
        };
        /**
         * See [[LockChain.addressToBytes]].
         */
        this.addressToBytes = (address) => {
            try {
                return bs58_1.default.decode(address);
            }
            catch (error) {
                try {
                    const [type, ...words] = bech32_1.default.decode(address).words;
                    return Buffer.concat([
                        Buffer.from([type]),
                        Buffer.from(bech32_1.default.fromWords(words)),
                    ]);
                }
                catch (internalError) {
                    throw new Error(`Unrecognized address format "${address}".`);
                }
            }
        };
        /**
         * See [[LockChain.bytesToAddress]].
         */
        this.bytesToAddress = (address) => {
            const words = bech32_1.default.toWords(address);
            return bech32_1.default.encode("", words);
        };
        /** @deprecated. Renamed to addressToBytes. */
        this.addressStringToBytes = this.addressToBytes;
        this.addressToString = (address) => address;
        /**
         * See [[LockChain.transactionID]].
         */
        this.transactionID = (transaction) => transaction.txHash;
        this.transactionIDFromRPCFormat = (txid, _txindex, reversed) => {
            // RenVM returns TXIDs in the correct byte direction, so they should be
            // reversed when converting to a string.
            // See https://learnmeabitcoin.com/technical/txid#why
            if (reversed) {
                // Reverse bytes.
                const bufferTxid = typeof txid === "string"
                    ? Buffer.from(utils_1.strip0x(txid), "hex")
                    : // Create new buffer because `reverse` is in-place.
                        Buffer.from(txid);
                return bufferTxid.reverse().toString("hex");
            }
            else {
                return typeof txid === "string" ? txid : txid.toString("hex");
            }
        };
        this.transactionFromRPCFormat = async (txid, txindex, reversed) => {
            const txidString = this.transactionIDFromRPCFormat(txid, txindex, reversed);
            return this.api.fetchUTXO(txidString, parseInt(txindex, 10));
        };
        /**
         * @deprecated Renamed to `transactionFromRPCFormat`.
         * Will be removed in 3.0.0.
         */
        this.transactionFromID = this.transactionFromRPCFormat;
        this.depositV1HashString = ({ transaction }) => {
            return `${utils_1.toBase64(utils_1.fromHex(transaction.txHash))}_${transaction.vOut}`;
        };
        this.transactionRPCFormat = (transaction, v2) => {
            const { txHash, vOut } = transaction;
            utils_1.assertType("string", { txHash });
            utils_1.assertType("number", { vOut });
            return {
                txid: v2
                    ? utils_1.fromHex(transaction.txHash).reverse()
                    : utils_1.fromHex(transaction.txHash),
                txindex: transaction.vOut.toFixed(),
            };
        };
        this.transactionRPCTxidFromID = (transactionID, v2) => v2 ? utils_1.fromHex(transactionID).reverse() : utils_1.fromHex(transactionID);
        /**
         * When burning, you can call `Bitcoin.Address("...")` to make the address
         * available to the burn params.
         *
         * @category Main
         */
        this.Address = (address) => {
            // Type validation
            utils_1.assertType("string", { address });
            this.burnPayloadGetter = (bytes) => bytes ? this.addressToBytes(address).toString("hex") : address;
            return this;
        };
        this.burnPayload = (burnPayloadConfig) => {
            return this.burnPayloadGetter
                ? this.burnPayloadGetter(burnPayloadConfig && burnPayloadConfig.bytes)
                : undefined;
        };
        this.toSats = (value) => new bignumber_js_1.default(value)
            .times(new bignumber_js_1.default(10).exponentiatedBy(8))
            .decimalPlaces(0)
            .toFixed();
        this.fromSats = (value) => new bignumber_js_1.default(value)
            .dividedBy(new bignumber_js_1.default(10).exponentiatedBy(8))
            .toFixed();
        this.chainNetwork = network;
    }
}
exports.BitcoinBaseChain = BitcoinBaseChain;
BitcoinBaseChain.chain = "Bitcoin";
// Asset
BitcoinBaseChain.asset = "BTC";
// Utils
BitcoinBaseChain.utils = {
    p2shPrefix: {},
    addressBufferToString: bs58_1.default.encode,
    addressIsValid: (_address, _network = "mainnet") => true,
    transactionIsValid: (_transaction, _network = "mainnet") => true,
    addressExplorerLink: (_address, _network = "mainnet") => undefined,
    transactionExplorerLink: (_tx, _network = "mainnet") => undefined,
    resolveChainNetwork: (network) => {
        if (network === "mainnet" ||
            network === "testnet" ||
            network === "regtest") {
            return network;
        }
        const renNetwork = interfaces_1.getRenNetworkDetails(network);
        return renNetwork.isTestnet ? "testnet" : "mainnet";
    },
};
const _ = BitcoinBaseChain;
