// @renproject/chains-ethereum/build/module/fantom.js
import { getRenNetworkDetails, RenNetwork, } from "@renproject/interfaces";
import { Callable, utilsWithChainNetwork } from "@renproject/utils";
import { EthereumClass } from "./ethereum";
import { StandardExplorer } from "./networks";
import { addressIsValid, transactionIsValid } from "./utils";
export const renFantomTestnet = {
    name: "Fantom Testnet",
    chain: "fantomTestnet",
    isTestnet: true,
    chainLabel: "Fantom Testnet",
    networkID: 0xfa2,
    addresses: {
        GatewayRegistry: "0x1207765B53697a046DCF4AE95bd4dE99ef9D3D3C",
        BasicAdapter: "0x07deB3917d234f787AEd86E0c88E829277D4a33b",
    },
    publicProvider: () => `https://rpc.testnet.fantom.network/`,
    explorer: StandardExplorer("https://testnet.ftmscan.com"),
    /** @deprecated Renamed to publicProvider. Will be removed in 3.0.0. */
    infura: "https://rpc.testnet.fantom.network/",
    /** @deprecated Renamed to explorer. Will be removed in 3.0.0. */
    etherscan: "https://testnet.ftmscan.com",
};
export const renFantomDevnet = Object.assign(Object.assign({}, renFantomTestnet), { addresses: {
        GatewayRegistry: "0xD881213F5ABF783d93220e6bD3Cc21706A8dc1fC",
        BasicAdapter: "0xD087b0540e172553c12DEEeCDEf3dFD21Ec02066",
    } });
export const renFantomMainnet = {
    name: "Fantom Mainnet",
    chain: "fantomMainnet",
    isTestnet: false,
    chainLabel: "Fantom Mainnet",
    networkID: 250,
    addresses: {
        GatewayRegistry: "0x21C482f153D0317fe85C60bE1F7fa079019fcEbD",
        BasicAdapter: "0xAC23817f7E9Ec7EB6B7889BDd2b50e04a44470c5",
    },
    publicProvider: () => `https://rpcapi.fantom.network`,
    explorer: StandardExplorer("https://ftmscan.com"),
    /** @deprecated Renamed to publicProvider. Will be removed in 3.0.0. */
    infura: "https://rpcapi.fantom.network",
    /** @deprecated Renamed to explorer. Will be removed in 3.0.0. */
    etherscan: "https://ftmscan.com",
};
export const FantomConfigMap = {
    [RenNetwork.Testnet]: renFantomTestnet,
    [RenNetwork.Mainnet]: renFantomMainnet,
    [RenNetwork.Devnet]: renFantomDevnet,
};
const resolveFantomNetwork = (renNetwork) => {
    if (!renNetwork) {
        return FantomConfigMap[RenNetwork.Mainnet];
    }
    if (renNetwork.addresses) {
        return renNetwork;
    }
    else {
        const details = getRenNetworkDetails(
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
        renNetwork);
        return details.isTestnet
            ? details.name === RenNetwork.Devnet
                ? renFantomDevnet
                : renFantomTestnet
            : renFantomMainnet;
    }
};
export class FantomClass extends EthereumClass {
    constructor(web3Provider, renNetwork, config = {}) {
        // To be compatible with the Ethereum chain class, the first parameter
        // is a web3Provider and the second the RenVM network.
        super(web3Provider, resolveFantomNetwork(renNetwork), config);
        this.chain = FantomClass.chain;
        this.name = FantomClass.chain;
        this.legacyName = undefined;
        this.configMap = FantomConfigMap;
        this.utils = utilsWithChainNetwork(FantomClass.utils, () => this.renNetworkDetails);
        this.initialize = (renNetwork) => {
            this.renNetworkDetails =
                this.renNetworkDetails ||
                    FantomConfigMap[getRenNetworkDetails(renNetwork).name];
            if (!this.renNetworkDetails) {
                throw new Error(`Unable to set ${this.name} network for RenVM network ${getRenNetworkDetails(renNetwork).name}. Please provide ${this.name} network details to ${this.name} constructor.`);
            }
            return this;
        };
    }
}
FantomClass.chain = "Fantom";
FantomClass.configMap = FantomConfigMap;
FantomClass.utils = {
    resolveChainNetwork: resolveFantomNetwork,
    addressIsValid,
    transactionIsValid,
    addressExplorerLink: (address, network) => `${(FantomClass.utils.resolveChainNetwork(network) ||
        renFantomMainnet).etherscan}/address/${address}`,
    transactionExplorerLink: (transaction, network) => `${(Fantom.utils.resolveChainNetwork(network) || renFantomMainnet)
        .etherscan}/tx/${transaction || ""}`,
};
// @dev Removes any static fields, except `utils`.
export const Fantom = Callable(FantomClass);
