// @renproject/chains-bitcoin/build/main/APIs/insight.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Insight = exports.InsightClass = void 0;
const utils_1 = require("@renproject/utils");
const axios_1 = __importDefault(require("axios"));
const https_1 = __importDefault(require("https"));
const API_1 = require("./API");
class InsightClass {
    constructor(url) {
        this.fetchUTXOs = async (address, confirmations = 0) => {
            const url = `${this.url}/addr/${address}/utxo`;
            const response = await axios_1.default.get(url, {
                // TODO: Remove when certificate is fixed.
                httpsAgent: new https_1.default.Agent({
                    rejectUnauthorized: false,
                }),
                timeout: API_1.DEFAULT_TIMEOUT,
            });
            const data = typeof response.data === "string"
                ? JSON.parse(response.data)
                : response.data;
            return (await Promise.all(data
                .filter((utxo) => confirmations === 0 ||
                utxo.confirmations >= confirmations)
                .map((utxo) => ({
                txHash: utxo.txid,
                amount: utils_1.isDefined(utxo.satoshis)
                    ? utxo.satoshis.toString()
                    : utils_1.isDefined(utxo.amount)
                        ? API_1.fixValue(utxo.amount, 8).toFixed()
                        : undefined,
                // script_hex: utxo.scriptPubKey,
                vOut: utxo.vout,
                confirmations: utxo.confirmations || 0,
            }))
                // If the amount is undefined, fetch the UTXO again.
                // This is due to the Digibyte explorer not returning
                // amounts correctly when fetching UTXOs.
                .map((utxo) => utils_1.isDefined(utxo.amount)
                ? utxo
                : this.fetchUTXO(utxo.txHash, utxo.vOut)))).sort(API_1.sortUTXOs);
        };
        this.fetchTXs = async (address, confirmations = 0) => {
            const url = `${this.url}/txs/?address=${address}`;
            const response = await axios_1.default.get(url, {
                // TODO: Remove when certificate is fixed.
                httpsAgent: new https_1.default.Agent({
                    rejectUnauthorized: false,
                }),
                timeout: API_1.DEFAULT_TIMEOUT,
            });
            const data = typeof response.data === "string"
                ? JSON.parse(response.data)
                : response.data;
            const received = [];
            for (const tx of data.txs) {
                for (let i = 0; i < tx.vout.length; i++) {
                    const vout = tx.vout[i];
                    if (vout.scriptPubKey.addresses.indexOf(address) >= 0) {
                        received.push({
                            txHash: tx.txid,
                            amount: API_1.fixValue(parseFloat(vout.value), 8).toFixed(),
                            vOut: i,
                            confirmations: tx.confirmations || 0,
                        });
                    }
                }
            }
            return received
                .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
                .sort(API_1.sortUTXOs);
        };
        this.fetchUTXO = async (txHash, vOut) => {
            const url = `${this.url}/tx/${txHash}`;
            const tx = (await axios_1.default.get(url, { timeout: API_1.DEFAULT_TIMEOUT })).data;
            return API_1.fixUTXO({
                txHash,
                amount: tx.vout[vOut].value.toString(),
                vOut,
                confirmations: tx.confirmations,
            }, 8);
        };
        this.broadcastTransaction = async (txHex) => {
            const url = `${this.url}/tx/send`;
            const response = await axios_1.default.post(url, { rawtx: txHex }, { timeout: API_1.DEFAULT_TIMEOUT });
            if (response.data.error) {
                throw new Error(response.data.error);
            }
            return response.data.txid;
        };
        this.url = url.replace(/\/$/, "");
    }
}
exports.InsightClass = InsightClass;
exports.Insight = utils_1.Callable(InsightClass);
